

# Introduction to Natural Language Processing (NLP) For Saltanat


---


![](https://snag.gy/uvESGH.jpg)

## What is Natural Language processing NLP ?
NLP in machine learning has many branches, suppose you wanted to predict the next word of a sentence or classify a book a review as negative or positive. These examples involve the data being in the form of words, depending on your task at hand you will need to represent your text data in a usable form so an algorithm can work.  

## What kind of data do we have at Yapily  and what part of NLP are we looking at?

At Yapily we look at transactional data, as you may know, this data is relatively short for each transaction and doesn't have any emotive meaning to it like a movie review (trying to class a reviews as negative or positive is a branch of NLP called Sentimental analysis). However, each transaction has a category (EATING_OUT, INCOME .. etc) and so at Yapily we are trying to build a model that will place an unknow transaction into the right category. For this to work we need to train a model on all the labelled data; so when that model sees a new transaction that has not been labelled we can place it in the right category at least with a high probability. For example, if the model sees TESCO then it will hopefully place that transaction in the Groceries category.





## Learning Objectives
- understanding how to extract features from the transaction data using the sckit learn package, this will set up the data in a format so the algorithm can work:
    - Count Vectorizer
    
    


## A Simple Example
---
Below I have created two transactions, the first (transaction_1) and the second is (transaction_2)

In [1]:
transaction_1= """
Sports Direct"""

transaction_2="""Tesco"""




print(transaction_1)
print()
print(transaction_2)


Sports Direct

Tesco


As you can see above I have these two transactions but i want them in a matrix so my algortihm can work, remember that for most algorithms we need the data to be in numeric form so we have to convert these two transactions into a numeric maxtrix and sklearn provides a great package to do this called the CountVectorizer from the sklearn.feature_extraction.text library


# What is the count Vectorizer?

The count vectorizer will count the amount of times a word has appeared in that paritcular document, note in our case document is transaction.

In [20]:
from sklearn.feature_extraction.text import CountVectorizer
import pandas as pd
import numpy as np


In [3]:
cvec=CountVectorizer()


Note here the count vectorizer is a class here and i have assigned it to the variable cvec. 


In [4]:
cvec.fit([transaction_1,transaction_2])

CountVectorizer(analyzer='word', binary=False, decode_error='strict',
        dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
        lowercase=True, max_df=1.0, max_features=None, min_df=1,
        ngram_range=(1, 1), preprocessor=None, stop_words=None,
        strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
        tokenizer=None, vocabulary=None)

Note above that .fit() is a functions that belongs to the CountVectorizer class, For more on class and functions using python https://www.programiz.com/python-programming/class

In [65]:
df = pd.DataFrame(cvec.transform([transaction_1,transaction_2]).toarray(),
                  columns=cvec.get_feature_names())




Only after fitting my CountVectorizer() over  can I transform the data, also note that cvec.transform([transaction_1,transaction_2]) will give a sparse matrix the toarray() function will convert that to normal 2 dimensional matrix which is put into a pandas dataframe df. For more on fit(),transform() and fit_trasform() functions check out https://towardsdatascience.com/fit-vs-transform-in-scikit-libraries-for-machine-learning-3c70e6300ded and for more on 
sparse matrix check out https://machinelearningmastery.com/sparse-matrices-for-machine-learning/

In [6]:
df

direct  sports  tesco
0       1       1      0
1       0       0      1

Note the above pandas data frame we can see that there are only two rows, the first row represent transaction_1 and the second row represents transaction_2. The columns are all the different words found in both the transactions, as it can be seen direct and sports is only found once in the first transaction and tesco is not found any times in the first transaction. Note in the second row tesco is found only once and direct and sports are not present in that transaction.

## So your first task Saltanat, with all the labelling you have done, lets make that into the right format so we can run algorithm on it later. This means using the CountVecotrizer over the transaction description.

The code below will read the data you have labelled and put it into a pandas dataframe


In [40]:
df=pd.read_csv("saltanat_labelled_batch_1_18_06_2020.csv")


Note Saltanat that you have labelled some transactions as "NAN" and this is not a category so we will have to drop those transactions from the data set. First, we have to convert those values to missing value so that makes it easier to drop.


In [41]:
df.labels.value_counts()

EATING_OUT                                                988
NAN                                                       954
SHOPPING                                                  576
GROCERIES                                                 486
OTHER                                                     465
FEES_&_CHARGES                                            344
TRANSPORT                                                 322
INVESTMENT_SAVE                                           249
ENTERTAINMENT                                             181
HOME                                                       94
PERSONAL_CARE                                              70
BILLS                                                      64
INSURANCE                                                  63
GENERAL                                                    48
TRAVEL_&_HOLIDAY                                           45
CHARITY_AND_GIFTS                                          21
EATING_O

Note above how the NAN value is at 954, what happens after I converted that string into a missing value

In [42]:
df.labels=df.labels.replace('NAN',np.nan)


The above code will replace "NAN" in the labels column with np.nan missing value, check now below what happens when you run value_counts. Note that np comes from the numpy library that was imported earlier.

In [43]:
df.labels.value_counts()



EATING_OUT                                                988
SHOPPING                                                  576
GROCERIES                                                 486
OTHER                                                     465
FEES_&_CHARGES                                            344
TRANSPORT                                                 322
INVESTMENT_SAVE                                           249
ENTERTAINMENT                                             181
HOME                                                       94
PERSONAL_CARE                                              70
BILLS                                                      64
INSURANCE                                                  63
GENERAL                                                    48
TRAVEL_&_HOLIDAY                                           45
CHARITY_AND_GIFTS                                          21
EATING_OUT                                                  7
TRANSOPR

Thats right the "NAN" labels are considered missing so pandas just doesn't count them. Now let's drop all the missing values in the data set before we Countvectorize it.

In [44]:
print(df.shape)

(4999, 23)


Note the shape of the data before you dropped the missing value.

In [49]:
df.dropna(subset=["labels"], inplace=True)

In [50]:
df.shape


(4045, 23)

Note that the shape of your data now its droped down to 4045

In [54]:
cvec.fit(df.anonymized_description)

CountVectorizer(analyzer='word', binary=False, decode_error='strict',
        dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
        lowercase=True, max_df=1.0, max_features=None, min_df=1,
        ngram_range=(1, 1), preprocessor=None, stop_words=None,
        strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
        tokenizer=None, vocabulary=None)

Above I have just put df.anonymized_description the column that i want to be vectorized, it therefore takes every transaction description into account in the data set.

In [58]:
transaction_matrix_data_frame=pd.DataFrame(cvec.transform(df.anonymized_description).toarray(),
                  columns=cvec.get_feature_names())

In [60]:
transaction_matrix_data_frame

000000  0006  0006051183  0010  0011462998  0012317699  0024427914  \
0          0     0           0     0           0           0           0   
1          0     0           0     0           0           0           0   
2          0     0           0     0           0           0           0   
3          0     0           0     0           0           0           0   
4          0     0           0     0           0           0           0   
5          0     0           0     0           0           0           0   
6          0     0           0     0           0           0           0   
7          0     0           0     0           0           0           0   
8          0     0           0     0           0           0           0   
9          0     0           0     0           0           0           0   
10         0     0           0     0           0           0           0   
11         0     0           0     0           0           0           0   
12         0     0           0     0           0           0           0   
13         0     0           0     0           0           0           0   
14         0     0           0     0           0           0           0   
15         0     0           0     0           0           0           0   
16         0     0           0     0           0           0           0   
17         0     0           0     0           0           0           0   
18         0     0           0     0           0           0           0   
19         0     0           0     0           0           0           0   
20         0     0           0     0           0           0           0   
21         0     0           0     0           0           0           0   
22         0     0           0     0           0           0           0   
23         0     0           0     0           0           0           0   
24         0     0           0     0           0           0           0   
25         0     0           0     0           0           0           0   
26         0     0           0     0           0           0           0   
27         0     0           0     0           0           0           0   
28         0     0           0     0           0           0           0   
29         0     0           0     0           0           0           0   
...      ...   ...         ...   ...         ...         ...         ...   
4015       0     0           0     0           0           0           0   
4016       0     0           0     0           0           0           0   
4017       0     0           0     0           0           0           0   
4018       0     0           0     0           0           0           0   
4019       0     0           0     0           0           0           0   
4020       0     0           0     0           0           0           0   
4021       0     0           0     0           0           0           0   
4022       0     0           0     0           0           0           0   
4023       0     0           0     0           0           0           0   
4024       0     0           0     0           0           0           0   
4025       0     0           0     0           0           0           0   
4026       0     0           0     0           0           0           0   
4027       0     0           0     0           0           0           0   
4028       0     0           0     0           0           0           0   
4029       0     0           0     0           0           0           0   
4030       0     0           0     0           0           0           0   
4031       0     0           0     0           0           0           0   
4032       0     0           0     0           0           0           0   
4033       0     0           0     0           0           0           0   
4034       0     0           0     0           0           0           0   
4035       0     0           0     0           0      

As you can see above, in total there are 1472 columns from the 4045 transactions this means that there are 1472 diffent words in all the transactions.Check out below what are the most popular words throughout all transaction. 

In [63]:
transaction_matrix_data_frame.sum().sort_values(ascending=False)

the             367
uk              289
london          231
co              207
com             206
sainsburys      202
change          202
save            202
fee             142
amazon          141
mkts            131
ltd             122
uber            121
www             107
tesco           106
ticket          103
holybelly        96
stores           92
amznmktplace     88
tfl              84
cp               82
gov              77
non              76
plum             72
int              69
nowtv            64
sodexo           63
first            63
coff             61
amazoamazon      60
               ... 
dyson             1
dvsa              1
pho               1
photo             1
dubai             1
payments          1
el                1
electr            1
event             1
fabfab            1
f3                1
walmer            1
experimental      1
papa              1
exmouth           1
ew                1
everyman          1
everyday          1
parkwood          1


## The first task Saltanant is to convert your batched 2 fully labelled data into the a dataframe (matrix) so that we can use it for an algorithm, by using CountVectorizer from sklearn.

## The Second task it to fiddle around with the different arguments in the Count Vectorizer , rerun your code above, but this time with the CountVectorizer( ngram_range=(2, 2) ). Check out https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html for the different argument you can use in the CountVectorizer.
